storage/src/bmt/mod.rs

```diff
@@ -0,0 +1,809 @@
+//! number of nodes, the last node is duplicated). The root of the tree is the digest of the top level (only
+//! ever one node).
+//!
+//! For example, given three leaves A, B, and C, the tree is constructed as follows:
+//!
+//! ```text
+//!     Level 2 (root):       [hash(hash(hash(0,A),hash(1,B)),hash(hash(2,C),hash(2,C)))]
+//!     Level 1:              [hash(hash(0,A),hash(1,B)),hash(hash(2,C),hash(2,C))]
+//!     Level 0 (leaves):     [hash(0,A),hash(1,B),hash(2,C)]
+//! ```
+//!
+//! A proof for a given leaf is generated by collecting the sibling at each level (from the leaf up to the root).
+//! An external process can then use this proof (with some trusted root) to verify that the leaf (at a fixed position)
+//! is part of the tree.
+//!
+//! # Example
+//!
+//! ```rust
+//! use commonware_storage::bmt::{Builder, Tree};
+//! use commonware_cryptography::{hash, Sha256, sha256::Digest};
+//!
+//! // Create transactions and compute their digests
+//! let txs = [b"tx1", b"tx2", b"tx3", b"tx4"];
+//! let digests: Vec<Digest> = txs.iter().map(|tx| hash(*tx)).collect();
+//!
+//! // Build a Merkle Tree from the digests
+//! let mut builder = Builder::<Sha256>::new(digests.len());
+//! for digest in &digests {
+//!    builder.add(digest);
+//! }
+//! let tree = builder.build().unwrap();
+//! let root = tree.root();
+//!
+//! // Generate a proof for leaf at index 1
+//! let mut hasher = Sha256::default();
+//! let proof = tree.proof(0).unwrap();
+//! assert!(proof.verify(&mut hasher, &digests[0], 0, &root));
+//! ```
+
+use bytes::Buf;
+use commonware_cryptography::{Array, Hasher};
+use commonware_utils::SizedSerialize;
+
+/// Constructor for a Binary Merkle Tree (BMT).
+pub struct Builder<H: Hasher> {
+    hasher: H,
+    leaves: Vec<H::Digest>,
+}
+
+impl<H: Hasher> Builder<H> {
+    /// Creates a new Binary Merkle Tree builder.
+    pub fn new(leaves: usize) -> Self {
+        Self {
+            hasher: H::new(),
+            leaves: Vec::with_capacity(leaves),
+        }
+    }
+
+    /// Adds a leaf to the Binary Merkle Tree.
+    ///
+    /// When added, the leaf is hashed with its position.
+    pub fn add(&mut self, leaf: &H::Digest) {
+        let position: u32 = self.leaves.len().try_into().expect("too many leaves");
+        self.hasher.update(&position.to_be_bytes());
+        self.hasher.update(leaf);
+        self.leaves.push(self.hasher.finalize());
+    }
+
+    /// Builds the Binary Merkle Tree.
+    pub fn build(self) -> Option<Tree<H>> {
+        Tree::new(self.hasher, self.leaves)
+    }
+}
+
+/// Constructed Binary Merkle Tree (BMT).
+#[derive(Clone, Debug)]
+pub struct Tree<H: Hasher> {
+    /// The digests at each level of the tree (from leaves to root).
+    levels: Vec<Vec<H::Digest>>,
+}
+
+impl<H: Hasher> Tree<H> {
+    /// Builds a Merkle Tree from a slice of position-hashed leaf digests.
+    ///
+    /// If `leaves` is empty, returns `None`.
```

Comment:
If there will be cases where we could end up needing to include a "root hash" from an empty set, it might be better to allow for empty merkle trees that have a canonical "empty root".

Reply:
Good call. Need to think about how to do that.

Reply:
You could use the default output of the hasher when you call finalize without having provided any data? (I'd guess it's just all zeroes but not really sure.)

Reply:
I was doing something dumber to get this. Great idea.